src/processor/operator/persistent/node_batch_insert.cpp

```diff
@@ -34,8 +37,35 @@ void NodeBatchInsertSharedState::initPKIndex(const ExecutionContext* context) {
 }
 
 void NodeBatchInsert::initGlobalStateInternal(ExecutionContext* context) {
-    const auto nodeSharedState = ku_dynamic_cast<NodeBatchInsertSharedState*>(sharedState.get());
+    auto clientContext = context->clientContext;
+    auto storageManager = clientContext->getStorageManager();
+    auto transaction = clientContext->getTransaction();
+
+    auto tableEntry = clientContext->getCatalog()->getTableCatalogEntry(transaction, tableName);
+    auto nodeTableEntry = tableEntry->ptrCast<NodeTableCatalogEntry>();
+    auto nodeTable = storageManager->getTable(nodeTableEntry->getTableID());
+
+    auto fTable = FactorizedTableUtils::getSingleStringColumnFTable(clientContext->getMemoryManager());
+
+    const auto& pkDefinition = nodeTableEntry->getPrimaryKeyDefinition();
+    auto pkColumnID = nodeTableEntry->getColumnID(pkDefinition.getName());
+
+    auto nodeSharedState = sharedState->ptrCast<NodeBatchInsertSharedState>();
+    nodeSharedState->table = nodeTable;
+    nodeSharedState->pkColumnID = pkColumnID;
+    nodeSharedState->pkType = pkDefinition.getType().copy();
     nodeSharedState->initPKIndex(context);
+
+    if (children[0]->getOperatorType() == PhysicalOperatorType::TABLE_FUNCTION_CALL) {
+        const auto call = children[0]->ptrCast<TableFunctionCall>();
+        nodeSharedState->tableFuncSharedState = call->getSharedState().get();
+    }
```

Comment:
You may not need to move this logic inside `NodeBatchInsert::initGlobalStateInternal`. Can you double check?

Reply:
Fixed with acc6017393f9fe85c42fbe7ccb02d426b2b954b4